storage/innobase/dict/dict0dict.cc

```diff
@@ -7450,27 +7450,93 @@ dberr_t dict_get_dictionary_info_by_id(ulint dict_id, char **name,
+static bool dict_is_mysql_plugin_space_encrypted(
+    space_id_t mysql_plugin_space) {
+  if (mysql_plugin_space == SYSTEM_TABLE_SPACE) {
+    return (srv_sys_space.is_encrypted());
+  }
+
+  space_id_t space_id = fil_space_get_id_by_name("mysql/plugin");
+  if (space_id == SPACE_UNKNOWN) {
+    return (false);
+  }
 
-    space_id_t space_id = fil_space_get_id_by_name("mysql/plugin");
-    if (space_id == SPACE_UNKNOWN) {
-      return (false);
-    }
+  const fil_space_t *space = fil_space_get(space_id);
+  ut_ad(space != nullptr);
+
+  if (space == nullptr) {
+    return false;
+  }
+  return FSP_FLAGS_GET_ENCRYPTION(space->flags);
+}
+
+/** Detect if innodb_encrypt_tables is set either to ON or FORCE
+@return true if innodb_encrypt_tables is equal to ON or FORCE */
+static bool dict_is_encrypt_tables_either_on_or_force() {
+  return srv_encrypt_tables == SRV_ENCRYPT_TABLES_ON ||
+         srv_encrypt_tables == SRV_ENCRYPT_TABLES_FORCE;
+}
 
-    fil_space_t *space = fil_space_get(space_id);
-    ut_ad(space != nullptr);
+/** Detect if mysql.ibd's page0 has encryption flag set.
+ *  The page 0 is read from disk.
+@param[in] filepath path of mysql.ibd
+@param[out] success true if no error
+@return true if encryption flag is set, false otherwise */
+static bool dict_mysql_ibd_page_0_has_encryption_flag_set(const char *filepath,
+                                                          bool &success) {
+  pfs_os_file_t file = os_file_create_simple_no_error_handling(
+      innodb_data_file_key, filepath, OS_FILE_OPEN, OS_FILE_READ_ONLY,
+      srv_read_only_mode, &success);
 
-    if (space == nullptr) {
-      return (false);
-    }
-    encrypt_mysql = FSP_FLAGS_GET_ENCRYPTION(space->flags);
+  if (!success) {
+    return (false);
+  }
+
+  /* Read the first page of the tablespace */
+  auto buf = ut_make_unique_ptr_nokey(2 * UNIV_PAGE_SIZE);
+  /* Align memory for file I/O if we might have O_DIRECT set */
+  byte *page = static_cast<byte *>(ut_align(buf.get(), UNIV_PAGE_SIZE));
+
+  ut_ad(page == page_align(page));
+
+  IORequest request(IORequest::READ);
+
+  dberr_t err = os_file_read_first_page(request, file, page, UNIV_PAGE_SIZE);
+
+  if (DB_SUCCESS != err) {
+    os_file_close(file);
+    success = false;
+    return (false);
   }
 
-  return (encrypt_mysql || srv_encrypt_tables == SRV_ENCRYPT_TABLES_ON ||
-          srv_encrypt_tables == SRV_ENCRYPT_TABLES_FORCE);
+  const ulint flags = fsp_header_get_flags(page);
+  os_file_close(file);
+  success = true;
+  return (FSP_FLAGS_GET_ENCRYPTION(flags));
+}
+
+bool dict_detect_encryption_of_mysql_ibd(dict_init_mode_t dict_init_mode,
+                                         space_id_t mysql_plugin_space,
+                                         const char *filepath,
+                                         bool &encrypt_mysql) {
+  bool success = false;
+  switch (dict_init_mode) {
+    case DICT_INIT_CREATE_FILES:
+      encrypt_mysql = dict_is_encrypt_tables_either_on_or_force();
```

Comment:
I see that you really like long names (function name that can go upto 60-80 characters).

Currently this handles only  ON or FORCE. So you made a choice to a "self describing" function which mentions value as part of function name. So if this function extends to handle below values,
SRV_ENCRYPT_TABLES_KEYRING_ON:
SRV_ENCRYPT_TABLES_KEYRING_FORCE:
SRV_ENCRYPT_TABLES_ONLINE_TO_KEYRING:
SRV_ENCRYPT_TABLES_ONLINE_TO_KEYRING_FORCE:

the function name should be:
dict_is_encrypt_tables_either_on_or_force_or_keyring_on_or_keyring_force_or_online_keyring_to_keyring_or_online_to_keyring_force() ? isn't it ;)

IMHO, function name should be precise and clear. This doesn't mean "Long explaining function names".

Reply:
I disagree, but this is code style and I will stay with longer names.

Reply:
Ok. Enjoy using them.

Other than that, why are you handling only ON and FORCE? what about the other enum values?